Reset animation on playback stop

KoBeWi opened the pull request:
Fixes #27499

KoBeWi committed:
Reset animation on playback stop

fire commented:
@akien-mga do you see any danger in merging this?

akien-mga commented:
> do you see any danger in merging this?

Yes, two things I'm not sure about:
- `seek(0, true)` calls `_animation_process(0)`, which does a whole lot of update logic. I'm not sure if it's good per se to have so much logic running when the user just wants to stop the animation from playing.
- More importantly, I think there are two conflicting use cases and we can't support both. Currently `stop(true)` will reset the position so that the next `play()` starts from 0, but it will not change the displayed frame back to frame 0. It's actually wanted in various situations if you want to stop the animation while staying on the stop frame, and when you `play()` again it should restart from 0. If we merge this PR, you'd have to pause the animation, seek to 0, and then play.

aaronfranke commented:
@KoBeWi Is this still desired? If so, it needs to be rebased on the latest master branch.

akien-mga commented:
Needs a rebase to fix CI. But note that this is potentially superseded by #44345, we need a design decision on how we want the API to be like for animation playback (both AnimationPlayer and AnimatedSprite).

From what I remember #44345 was discussed in a PR review meeting and had mostly consensus, so we might want to re-review that and possibly close this PR while merging #44345.

golddotasksquestions commented:
> More importantly, I think there are two conflicting use cases and we can't support both. Currently stop(true) will reset the position so that the next play() starts from 0, but it will not change the displayed frame back to frame 0. **It's actually wanted in various situations if you want to stop the animation while staying on the stop frame, and when you play() again it should restart from 0.** If we merge this PR, you'd have to pause the animation, seek to 0, and then play.

@akien-mga , no, this is never wanted. This is the dirtiest behaviour I can imagine. You are mixing to very different commands here: Pause and Stop. Pause, pauses (timestamp stays where it is) stop stops playback (resets timestamp).
Godot is the only thing in existence that does this weird mixed behaviour! Why, really?
If a user wants to pause the playback, we need a .pause() for them. If they want to stop playback (It means reset timestamp to 0), it should do just that immediately. Stop is not Pause and Pause is not stop. They have two very well know icons: ⏹ ⏸
This is the behaviour in literally ever music, video or animation playing device (analog or digital) I ever got my sticky sausage fingers on since early childhood. Please let's _not_ come up with totally unique novel solutions for worldwide incredibly well established and solved UI flows.

KoBeWi commented:
Rebased. The animation process code is so complex that I couldn't think of how to reasonably implement "internal seeking", so I just added a bool called `is_stopping`. Hope it's acceptable.

TokageItLab commented:
Playback avoidance is needed not only in TYPE_METHOD but also in TYPE_AUDIO and TYPE_ANIMATION. So, it is better to do `continue;` at the beginning of those case statements rather than propagate it to `can_call`.

KoBeWi commented:
I'm already stopping TYPE_AUDIO, forgot about TYPE_ANIMATION.

TokageItLab commented on scene/animation/animation_player.cpp:
```diff
@@ -748,7 +748,7 @@ void AnimationPlayer::_animation_process_animation(AnimationData *p_anim, double
 					continue;
 				}
-				if (!p_is_current) {
+				if (!p_is_current || is_stopping) {
```
Nit pick for consistency:
```cpp
			case Animation::TYPE_METHOD: {
				if (!nc->node || is_stopping) {
					continue;
				}
```
However, it is doubtful what this brake exists for, so it may be done as follows.
```cpp
			case Animation::TYPE_METHOD: {
				if (!nc->node || is_stopping || !p_is_current) {
					continue;
				}
```

TokageItLab reviewed (changes requested):
The reset option should be propagated to the SubAnimationPlayer of the AnimationPlaybackTrack.
```cpp
void AnimationPlayer::_stop_playing_caches() {
	for (TrackNodeCache *E : playing_caches) {
		if (E->node && E->audio_playing) {
			E->node->call(SNAME("stop"));
		}
		if (E->node && E->animation_playing) {
			AnimationPlayer *player = Object::cast_to<AnimationPlayer>(E->node);
			if (!player) {
				continue;
			}
			player->stop();
		}
	}

	playing_caches.clear();
}
```
So, it should be fixed to like this:
```cpp
void AnimationPlayer::_stop_playing_caches(bool p_reset) {
~~
			if (p_reset) {
				player->stop();
			} else {
				player->pause();
			}
~~
}
```

TokageItLab commented:
Tested, looks fine except for the above issue. I sent a PR to apply the pause() and stop() separately to AnimationPlayerEditor as #71321. I will rebase that PR as soon as this PR is merged first.

TokageItLab reviewed (approved):
LGTM. I agree with this change, also it is consistent with tween.

akien-mga commented:
Thanks!

LakshayaG73 commented:
Hi. Does this fix imply that property values tracks will have their 0 time values reapplied when  an animation is stopped?

How is that desired behavior?
